ksqldb-metastore/src/main/java/io/confluent/ksql/metastore/MetaStoreImpl.java

```diff
@@ -295,25 +277,45 @@ public boolean deleteType(final String name) {
   private static final class SourceInfo {
 
     private final DataSource source;
-    private final ReferentialIntegrityTableEntry referentialIntegrity;
+    private final Set<SourceName> linkedSources = new ConcurrentHashSet<>();
+    private final Set<SourceName> referentialSources = new ConcurrentHashSet<>();
 
     private SourceInfo(
         final DataSource source
     ) {
       this.source = Objects.requireNonNull(source, "source");
-      this.referentialIntegrity = new ReferentialIntegrityTableEntry();
     }
 
     private SourceInfo(
         final DataSource source,
-        final ReferentialIntegrityTableEntry referentialIntegrity
+        final Set<SourceName> linkedSources,
+        final Set<SourceName> referentialSources
     ) {
       this.source = Objects.requireNonNull(source, "source");
-      this.referentialIntegrity = referentialIntegrity.copy();
+      this.linkedSources.addAll(
+          Objects.requireNonNull(linkedSources, "linkedSources"));
+      this.referentialSources.addAll(
+          Objects.requireNonNull(referentialSources, "referentialSources"));
     }
 
     public SourceInfo copy() {
-      return new SourceInfo(source, referentialIntegrity);
+      return new SourceInfo(source, linkedSources, referentialSources);
```

Comment:
```suggestion
      return copyWith(source);
```

Reply:
I ended up removing this `copyWith`. You made a comment in the `putSource` method that you were confusing, and I changed it to avoid calling this copy.